File: commons/src/main/java/l2s/commons/net/netty/ReceivablePacket.java
package l2s.commons.net.netty;

import io.netty.buffer.ByteBuf;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.nio.charset.StandardCharsets;

public abstract class ReceivablePacket<T> implements Runnable {
    protected static final Logger LOGGER = LogManager.getLogger(ReceivablePacket.class);

    private final T client;
    private final ByteBuf byteBuf;

    public ReceivablePacket(T client, ByteBuf byteBuf) {
        this.client = client;
        this.byteBuf = byteBuf;
    }

    public boolean isReadable() {
        return byteBuf.isReadable();
    }

    protected String readString(ByteBuf byteBuf) {
        var length = byteBuf.readShort();
        return byteBuf.readCharSequence(length, StandardCharsets.UTF_8).toString();
    }

    @Override
    public final void run() {
        try {
            runImpl(client);
        } catch (Exception e) {
            LOGGER.error("", e);
        }
    }

    protected abstract void runImpl(T client);
}

File: commons/src/main/java/l2s/commons/net/netty/ReceivablePacket.kt
package l2s.commons.net.netty

import com.google.common.flogger.FluentLogger
import io.netty.buffer.ByteBuf
import java.nio.charset.StandardCharsets

/**
 * @author Java-man
 */
abstract class ReceivablePacket<T>(private val client: T, private val byteBuf: ByteBuf) : Runnable {

    val isReadable: Boolean
        get() = byteBuf.isReadable

    protected fun readString(byteBuf: ByteBuf): String {
        val length = byteBuf.readShort()
        return byteBuf.readCharSequence(length.toInt(), StandardCharsets.UTF_8).toString()
    }

    override fun run() {
        try {
            runImpl(client)
        } catch (e: Exception) {
            LOGGER.atSevere().withStackTrace(com.google.common.flogger.StackSize.FULL).withCause(e).log("Can't read packet.")
        }
    }

    protected abstract fun runImpl(client: T)

    companion object {

        private val LOGGER = FluentLogger.forEnclosingClass()

    }

}